From 1c18818cf993cd41fb15630cb786c65184aaa7ca Mon Sep 17 00:00:00 2001
From: t <tristan75015@gmail.com>
Date: Tue, 6 Oct 2026 19:45:27 +0200
Subject: [PATCH] TFEL, MFront: error reporting in host and device code,
 portable test laws

- `tfel::fail` throws an exception of the given type on the host and
  returns `false` in device code, where exceptions are not supported. The
  calling code block returns this value.
- `TinyNonLinearSolverBase::abortResolution` makes the non linear solver
  fail. Returning `false` from the evaluation of the residual only reduces
  the Newton correction, whereas an exception aborts the resolution.
- `abortIntegration` (Runge-Kutta DSLs) makes the integration fail.
  Returning `false` from `@Derivative` only reduces the time step.

30 test behaviours and models of `mfront/tests` used exceptions, standard
streams or messages built at runtime in their code blocks. They now use
`tfel::fail`, and `abortResolution` when the error is reported during the
resolution of the implicit system. The standard streams and the messages
built at runtime are only used on the host. The behaviour on the host is
unchanged: the same exceptions, with the same messages, are thrown.
---
 .../generic-parallel-behaviour-interface.md   | 41 ++++++++++++++++---
 .../TinyNonLinearSolverBase.hxx               | 19 +++++++++
 include/TFEL/Raise.hxx                        | 24 +++++++++++
 include/TFEL/Raise.ixx                        |  9 ++++
 mfront/src/RungeKuttaCodeGeneratorBase.cxx    | 19 ++++++++-
 mfront/tests/behaviours/AxialGrowth3.mfront   |  2 +-
 .../CompressibleNeoHookeanBehaviour.mfront    |  3 +-
 ...initeRotationSmallStrainElasticity2.mfront |  6 ++-
 .../FiniteStrainSingleCrystal2.mfront         |  2 +-
 .../behaviours/GeneralizedMaxwell.mfront      |  4 +-
 mfront/tests/behaviours/Gurson.mfront         |  3 +-
 .../tests/behaviours/ImplicitNorton3.mfront   |  2 +-
 .../ImplicitSimoMieheElastoPlasticity.mfront  |  2 +-
 ...IncompressibleMooneyRivlinBehaviour.mfront |  3 +-
 ...ePlaneStressGornetDesmoratBehaviour.mfront |  9 ++--
 ...blePlaneStressMooneyRivlinBehaviour.mfront |  3 +-
 .../LogarithmicStrainElasticity2.mfront       |  4 ++
 ...LogarithmicStrainStandardElasticity.mfront |  2 +-
 ...ogarithmicStrainStandardElasticity2.mfront |  2 +-
 ...ogarithmicStrainStandardElasticity3.mfront |  2 +-
 ...ogarithmicStrainStandardElasticity4.mfront |  2 +-
 .../tests/behaviours/MonoCrystal_DD_CC.mfront | 14 +++++--
 .../MonoCrystal_DD_CC_Irradiation.mfront      | 14 +++++--
 ...DD_CC_Irradiation_NumericalJacobian.mfront | 18 ++++++--
 ...onoCrystal_DD_CC_LevenbergMarquardt.mfront | 14 +++++--
 ...MonoCrystal_DD_CC_NumericalJacobian.mfront | 18 ++++++--
 .../behaviours/OrthotropicElastic6.mfront     |  2 +
 mfront/tests/behaviours/ParameterTest.mfront  | 14 ++++---
 mfront/tests/behaviours/ParameterTest2.mfront | 14 ++++---
 .../SaintVenantKirchhoffElasticity2.mfront    |  4 +-
 ...tKirchhoffElasticityWithSubStepping.mfront |  2 +-
 .../TestComputeElasticPrediction.mfront       |  4 +-
 .../TestComputeElasticPrediction2.mfront      |  4 +-
 ...TemperatureUpdateOnSubSteppingCheck.mfront |  4 +-
 ...C_IrradiationSwellingModel_GoFaster.mfront |  4 ++
 35 files changed, 228 insertions(+), 65 deletions(-)

diff --git a/docs/web/generic-parallel-behaviour-interface.md b/docs/web/generic-parallel-behaviour-interface.md
index 33552035a..5242ba1b4 100644
--- a/docs/web/generic-parallel-behaviour-interface.md
+++ b/docs/web/generic-parallel-behaviour-interface.md
@@ -522,11 +522,42 @@ virtual functions are not supported in device code by SYCL.
 
 The following rules apply to the code blocks written by the user:
 
-- exceptions are not supported in device code: return `false` (in
-  `@InitLocalVariables`, `@Integrator` of the default DSL, etc.) or
-  `FAILURE` instead of throwing an exception. The `raise` and `raise_if`
-  functions can't be used either.
-- the standard streams (`std::cout`) can't be used.
+- exceptions are not supported in device code. The `tfel::fail` function
+  reports an error in host and device code: on the host, it throws an
+  exception of the given type, as before; in device code, it returns
+  `false`, which must be returned by the code block:
+
+  ~~~~{.cxx}
+  if (q3 > q1) {
+    return tfel::fail<std::runtime_error>("q3 is greater than q1");
+  }
+  ~~~~
+
+  The `raise` and `raise_if` functions can't be used in device code.
+- returning `false` from the `@Integrator` code block of the implicit DSLs
+  (or from `@Derivative` in the Runge-Kutta DSLs) does not make the
+  integration fail: the correction of the Newton algorithm (or the time
+  step) is reduced. An exception, on the contrary, makes the integration
+  fail. The same result is obtained in device code by calling
+  `this->abortResolution()` (implicit DSLs) or `this->abortIntegration()`
+  (Runge-Kutta DSLs) before `tfel::fail`:
+
+  ~~~~{.cxx}
+  if (omega < 0) {
+    this->abortResolution();
+    return tfel::fail<std::runtime_error>("negative value of omega");
+  }
+  ~~~~
+- the standard streams (`std::cout`) can't be used in device code. They
+  can be kept on the host:
+
+  ~~~~{.cxx}
+  #ifndef TFEL_DEVICE_CODE
+  std::cout << "sig: " << sig << '\n';
+  #endif /* TFEL_DEVICE_CODE */
+  ~~~~
+- error messages built at runtime (`std::ostringstream`, `std::to_string`)
+  are only available on the host, in the same way.
 - function scope static variables must be constant expressions. Constant
   objects built at compile time must be accessed through
   `tfel::utilities::getConstantObject` (`nvcc` does not initialize the
diff --git a/include/TFEL/Math/NonLinearSolvers/TinyNonLinearSolverBase.hxx b/include/TFEL/Math/NonLinearSolvers/TinyNonLinearSolverBase.hxx
index 0db376a18..eda54a08e 100644
--- a/include/TFEL/Math/NonLinearSolvers/TinyNonLinearSolverBase.hxx
+++ b/include/TFEL/Math/NonLinearSolvers/TinyNonLinearSolverBase.hxx
@@ -236,6 +236,25 @@ namespace tfel::math {
      */
     TFEL_HOST_DEVICE constexpr void reportStandardIteration(
         const NumericType) const noexcept {}
+    /*!
+     * \brief abort the resolution: the solver reports a failure as soon as
+     * the current evaluation of the residual or of the correction fails.
+     * This allows to abort the resolution in device code, where exceptions
+     * are not supported (see `tfel::fail`):
+     *
+     * \code{.cpp}
+     * if (omega < 0) {
+     *   this->abortResolution();
+     *   return tfel::fail("negative value of omega");
+     * }
+     * \endcode
+     *
+     * On the host, `tfel::fail` throws an exception, which aborts the
+     * resolution anyway.
+     */
+    TFEL_HOST_DEVICE constexpr void abortResolution() noexcept {
+      this->iter = this->iterMax;
+    }
     //! \brief criterion value
     NumericType epsilon;
     /*!
diff --git a/include/TFEL/Raise.hxx b/include/TFEL/Raise.hxx
index 0181909b1..5dad96c09 100644
--- a/include/TFEL/Raise.hxx
+++ b/include/TFEL/Raise.hxx
@@ -66,6 +66,30 @@ namespace tfel {
   template <typename Exception = std::runtime_error, typename... Args>
   TFEL_VISIBILITY_LOCAL TFEL_INLINE void raise_if(const bool, Args&&...);
 
+  /*!
+   * \brief report a failure in a function returning a boolean, or a type
+   * which can be built from a boolean, in host and device code:
+   *
+   * - on the host, an exception of the given type is thrown;
+   * - in device code, where exceptions are not supported, `false` is
+   *   returned. The caller shall return this value.
+   *
+   * \code{.cpp}
+   * if (q3 > q1) {
+   *   return tfel::fail<std::runtime_error>("q3 is greater than q1");
+   * }
+   * \endcode
+   *
+   * \tparam Exception: type of the exception thrown on the host.
+   * \tparam Args: type of the arguments passed to the exception'
+   * constructor.
+   * \param[in] a: arguments passed to the exception' constructor, which
+   * are only used on the host.
+   */
+  template <typename Exception = std::runtime_error, typename... Args>
+  [[nodiscard]] TFEL_VISIBILITY_LOCAL TFEL_HOST_DEVICE TFEL_INLINE bool fail(
+      Args&&...);
+
 }  // end of namespace tfel
 
 #include "TFEL/Raise.ixx"
diff --git a/include/TFEL/Raise.ixx b/include/TFEL/Raise.ixx
index 80d45c3f6..c0d717e01 100644
--- a/include/TFEL/Raise.ixx
+++ b/include/TFEL/Raise.ixx
@@ -42,6 +42,15 @@ namespace tfel {
     }
   }  // end of raise
 
+  template <typename Exception, typename... Args>
+  TFEL_HOST_DEVICE bool fail([[maybe_unused]] Args&&... a) {
+#ifdef TFEL_DEVICE_CODE
+    return false;
+#else  /* TFEL_DEVICE_CODE */
+    raise<Exception>(std::forward<Args>(a)...);
+#endif /* TFEL_DEVICE_CODE */
+  }  // end of fail
+
 }  // end of namespace tfel
 
 #endif /* LIB_TFEL_RAISE_IXX */
diff --git a/mfront/src/RungeKuttaCodeGeneratorBase.cxx b/mfront/src/RungeKuttaCodeGeneratorBase.cxx
index 41b32885c..0f8243e18 100644
--- a/mfront/src/RungeKuttaCodeGeneratorBase.cxx
+++ b/mfront/src/RungeKuttaCodeGeneratorBase.cxx
@@ -341,14 +341,31 @@ namespace mfront {
          << "} // end of " << this->bd.getClassName()
          << "::computeFinalThermodynamicForces\n\n";
     }
+    // once the integration is aborted, all the evaluations of the
+    // derivatives fail, so that the time step is reduced until the
+    // integration fails
     os << "[[nodiscard]] TFEL_HOST_DEVICE bool\ncomputeDerivative(){\n"
        << "using namespace std;\n"
-       << "using namespace tfel::math;\n";
+       << "using namespace tfel::math;\n"
+       << "if(this->mfront_aborted_integration){\n"
+       << "return false;\n"
+       << "}\n";
     writeMaterialLaws(os, this->bd.getMaterialLaws());
     os << this->bd.getCode(h, BehaviourData::ComputeDerivative) << '\n'
        << "return true;\n"
        << "} // end of " << this->bd.getClassName()
        << "::computeDerivative\n\n";
+    os << "/*!\n"
+       << " * \\brief abort the integration: the integration fails,\n"
+       << " * whereas a failure of the evaluation of the derivatives only\n"
+       << " * reduces the time step. This allows to report errors in device\n"
+       << " * code, where exceptions are not supported (see `tfel::fail`).\n"
+       << " */\n"
+       << "TFEL_HOST_DEVICE void abortIntegration() noexcept {\n"
+       << "this->mfront_aborted_integration = true;\n"
+       << "}\n\n"
+       << "//! \\brief boolean stating if the integration was aborted\n"
+       << "bool mfront_aborted_integration = false;\n\n";
   }  // end of writeBehaviourParserSpecificMembers
 
   void RungeKuttaCodeGeneratorBase::writeBehaviourUpdateStateVariables(
diff --git a/mfront/tests/behaviours/AxialGrowth3.mfront b/mfront/tests/behaviours/AxialGrowth3.mfront
index d9953d45b..706d6961f 100644
--- a/mfront/tests/behaviours/AxialGrowth3.mfront
+++ b/mfront/tests/behaviours/AxialGrowth3.mfront
@@ -22,7 +22,7 @@ ag.setGlossaryName("AxialGrowth");
 /* Initialize Lame coefficients */
 @InitLocalVariables{
   if(dt>100){
-    throw(DivergenceException());
+    return tfel::fail<DivergenceException>();
   }
 } // end of @InitLocalVars
 
diff --git a/mfront/tests/behaviours/CompressibleNeoHookeanBehaviour.mfront b/mfront/tests/behaviours/CompressibleNeoHookeanBehaviour.mfront
index 9f525a975..9bf314961 100644
--- a/mfront/tests/behaviours/CompressibleNeoHookeanBehaviour.mfront
+++ b/mfront/tests/behaviours/CompressibleNeoHookeanBehaviour.mfront
@@ -13,7 +13,6 @@
   const stress eg = 2*C1*pow(J,real(5)/real(3));
   sig = eg*b+pr*Stensor::Id();
   if(computeTangentOperator_){
-    string msg("tangent operator not yet available");
-    throw(runtime_error(msg));
+    return tfel::fail<runtime_error>("tangent operator not yet available");
   }
 }
diff --git a/mfront/tests/behaviours/FiniteRotationSmallStrainElasticity2.mfront b/mfront/tests/behaviours/FiniteRotationSmallStrainElasticity2.mfront
index 3acb657c4..12a8495d4 100644
--- a/mfront/tests/behaviours/FiniteRotationSmallStrainElasticity2.mfront
+++ b/mfront/tests/behaviours/FiniteRotationSmallStrainElasticity2.mfront
@@ -33,11 +33,13 @@ ps.setEntryName("PreviousStress");
   lambda = computeLambda(young,nu);
   mu = computeMu(young,nu);
   if(sqrt((sig-ps)|(sig-ps))>10*young*numeric_limits<stress>::epsilon()){
+#ifndef TFEL_DEVICE_CODE
     cout << "ps  : " << ps << endl;
     cout << "sig : " << sig << endl;
     cout << "nr  : " << sqrt((sig-ps)|(sig-ps)) << endl;
-    throw(runtime_error("invalid stress value at the beginning "
-			"of the time step"));
+#endif /* TFEL_DEVICE_CODE */
+    return tfel::fail<runtime_error>("invalid stress value at the beginning "
+                                     "of the time step");
   }
 }
 
diff --git a/mfront/tests/behaviours/FiniteStrainSingleCrystal2.mfront b/mfront/tests/behaviours/FiniteStrainSingleCrystal2.mfront
index bd5514750..7f3e4891b 100644
--- a/mfront/tests/behaviours/FiniteStrainSingleCrystal2.mfront
+++ b/mfront/tests/behaviours/FiniteStrainSingleCrystal2.mfront
@@ -93,7 +93,7 @@
 
 @InitLocalVariables{
   if(dt>1.e-5){
-    throw(DivergenceException());
+    return tfel::fail<DivergenceException>();
   }
   // increment du gradient de la transformation
   const Tensor dF = F1*invert(F0);
diff --git a/mfront/tests/behaviours/GeneralizedMaxwell.mfront b/mfront/tests/behaviours/GeneralizedMaxwell.mfront
index 48b7e407d..acae8dbb5 100644
--- a/mfront/tests/behaviours/GeneralizedMaxwell.mfront
+++ b/mfront/tests/behaviours/GeneralizedMaxwell.mfront
@@ -46,7 +46,7 @@ sej.setEntryName("ViscoelasticStress");
     }
   }
   if(computeTangentOperator_){
-    string msg("GenMaxwell : no tangent operator available");
-    throw(runtime_error(msg));
+    return tfel::fail<runtime_error>(
+        "GenMaxwell : no tangent operator available");
   }
 }
diff --git a/mfront/tests/behaviours/Gurson.mfront b/mfront/tests/behaviours/Gurson.mfront
index 941d57ec6..291425d0d 100644
--- a/mfront/tests/behaviours/Gurson.mfront
+++ b/mfront/tests/behaviours/Gurson.mfront
@@ -59,8 +59,7 @@ q3.setEntryName("ThirdGursonParameter");
   delta = ((ffb-fc)/(fmax-fc));
   const real feq = (f<fc ? f : fc+delta*(f-fc));
   if(q3>q1){
-    string msg("gurson : q3 is greater than q1 !");
-    throw(runtime_error(msg));
+    return tfel::fail<runtime_error>("gurson : q3 is greater than q1 !");
   }
   flim   = min(fc+delta*(fmax-fc),0.95*(q1-sqrt(q1*q1-q3*q3))/(q3*q3));
   broken = (feq>=flim);
diff --git a/mfront/tests/behaviours/ImplicitNorton3.mfront b/mfront/tests/behaviours/ImplicitNorton3.mfront
index 74507ccbb..9f31a3014 100644
--- a/mfront/tests/behaviours/ImplicitNorton3.mfront
+++ b/mfront/tests/behaviours/ImplicitNorton3.mfront
@@ -49,7 +49,7 @@ p.setGlossaryName("EquivalentViscoplasticStrain");
   lambda = computeLambda(young,nu);
   mu = computeMu(young,nu);
   if(dt>0.1){
-    throw(DivergenceException());
+    return tfel::fail<DivergenceException>();
   }
 } // end of @InitLocalVars
 
diff --git a/mfront/tests/behaviours/ImplicitSimoMieheElastoPlasticity.mfront b/mfront/tests/behaviours/ImplicitSimoMieheElastoPlasticity.mfront
index c3ac62ae7..750c3ba7e 100644
--- a/mfront/tests/behaviours/ImplicitSimoMieheElastoPlasticity.mfront
+++ b/mfront/tests/behaviours/ImplicitSimoMieheElastoPlasticity.mfront
@@ -132,7 +132,7 @@ Tref.setDefaultValue(293.15);
       x = x2;
     }
   } else {
-    throw(runtime_error("CubicRoots::exe : failed"));
+    return tfel::fail<runtime_error>("CubicRoots::exe : failed");
   }
   // elastic strain
   be = x*StrainStensor::Id()+dev_tK/mu;
diff --git a/mfront/tests/behaviours/IncompressibleMooneyRivlinBehaviour.mfront b/mfront/tests/behaviours/IncompressibleMooneyRivlinBehaviour.mfront
index 3b329de0f..2bb40964c 100644
--- a/mfront/tests/behaviours/IncompressibleMooneyRivlinBehaviour.mfront
+++ b/mfront/tests/behaviours/IncompressibleMooneyRivlinBehaviour.mfront
@@ -28,8 +28,7 @@
   sig = eg1*c-eg2*c2+pr*Stensor::Id();
   // no tangent operator yet
   if(computeTangentOperator_){
-    string msg("tangent operator not yet available");
-    throw(runtime_error(msg));
+    return tfel::fail<runtime_error>("tangent operator not yet available");
   }
 }
 
diff --git a/mfront/tests/behaviours/IncompressiblePlaneStressGornetDesmoratBehaviour.mfront b/mfront/tests/behaviours/IncompressiblePlaneStressGornetDesmoratBehaviour.mfront
index fea4c1912..57b7cc928 100644
--- a/mfront/tests/behaviours/IncompressiblePlaneStressGornetDesmoratBehaviour.mfront
+++ b/mfront/tests/behaviours/IncompressiblePlaneStressGornetDesmoratBehaviour.mfront
@@ -26,9 +26,13 @@
   const real i1 = trace(c);
   const real i2= (i1*i1-c(0)*c(0)-c(1)*c(1)-c(2)*c(2)-c(3)*c(3))/2;
   if(!(i2>0)){
+#ifndef TFEL_DEVICE_CODE
     ostringstream msg;
     msg << "invalid i2 value (" << i2 << ")";
-    throw(runtime_error(msg.str()));    
+    throw(runtime_error(msg.str()));
+#else  /* TFEL_DEVICE_CODE */
+    return false;
+#endif /* TFEL_DEVICE_CODE */
   }
   const real dw_di1 = H1*exp(H3*(i1-3)*(i1-3));
   const real dw_di2 = 3*H2/sqrt(i2);
@@ -42,7 +46,6 @@
   sig(2) = 0.;
   // no tangent operator yet
   if(computeTangentOperator_){
-    string msg("tangent operator not yet available");
-    throw(runtime_error(msg));
+    return tfel::fail<runtime_error>("tangent operator not yet available");
   }
 }
diff --git a/mfront/tests/behaviours/IncompressiblePlaneStressMooneyRivlinBehaviour.mfront b/mfront/tests/behaviours/IncompressiblePlaneStressMooneyRivlinBehaviour.mfront
index b573d752f..bb6ead2e0 100644
--- a/mfront/tests/behaviours/IncompressiblePlaneStressMooneyRivlinBehaviour.mfront
+++ b/mfront/tests/behaviours/IncompressiblePlaneStressMooneyRivlinBehaviour.mfront
@@ -33,7 +33,6 @@
   sig(2) = 0.;
   // no tangent operator yet
   if(computeTangentOperator_){
-    string msg("tangent operator not yet available");
-    throw(runtime_error(msg));
+    return tfel::fail<runtime_error>("tangent operator not yet available");
   }
 }
diff --git a/mfront/tests/behaviours/LogarithmicStrainElasticity2.mfront b/mfront/tests/behaviours/LogarithmicStrainElasticity2.mfront
index 8cfa2e3a8..24d660ef2 100644
--- a/mfront/tests/behaviours/LogarithmicStrainElasticity2.mfront
+++ b/mfront/tests/behaviours/LogarithmicStrainElasticity2.mfront
@@ -43,9 +43,13 @@ rp.setEntryName("RelativePrecision");
   const stress s   = sqrt(sig|sig);
   const stress eps = rp*s*numeric_limits<stress>::epsilon();
   if(ds>eps){
+#ifndef TFEL_DEVICE_CODE
     throw(runtime_error("invalid stress value at "
 			"the beginning of the time step "
 			"("+to_string(ds)+">"+to_string(eps)+")"));
+#else  /* TFEL_DEVICE_CODE */
+    return false;
+#endif /* TFEL_DEVICE_CODE */
   }
 }
 
diff --git a/mfront/tests/behaviours/LogarithmicStrainStandardElasticity.mfront b/mfront/tests/behaviours/LogarithmicStrainStandardElasticity.mfront
index 66dfe112f..e62802743 100644
--- a/mfront/tests/behaviours/LogarithmicStrainStandardElasticity.mfront
+++ b/mfront/tests/behaviours/LogarithmicStrainStandardElasticity.mfront
@@ -32,6 +32,6 @@
 @UpdateAuxiliaryStateVariables {
   const auto ds = eval(theta * sig + (1 - theta) * sig0 - sigel);
   if (sqrt(ds | ds) > 10 * epsilon * young) {
-    tfel::raise("invalid stress prediction");
+    return tfel::fail("invalid stress prediction");
   }
 }
\ No newline at end of file
diff --git a/mfront/tests/behaviours/LogarithmicStrainStandardElasticity2.mfront b/mfront/tests/behaviours/LogarithmicStrainStandardElasticity2.mfront
index ab8fb5599..1b9537e6e 100644
--- a/mfront/tests/behaviours/LogarithmicStrainStandardElasticity2.mfront
+++ b/mfront/tests/behaviours/LogarithmicStrainStandardElasticity2.mfront
@@ -38,6 +38,6 @@
 @UpdateAuxiliaryStateVariables {
   const auto ds = eval(theta * sig + (1 - theta) * sig0 - sigel);
   if (sqrt(ds | ds) > 10 * epsilon * young) {
-    tfel::raise("invalid stress prediction");
+    return tfel::fail("invalid stress prediction");
   }
 }
\ No newline at end of file
diff --git a/mfront/tests/behaviours/LogarithmicStrainStandardElasticity3.mfront b/mfront/tests/behaviours/LogarithmicStrainStandardElasticity3.mfront
index 6bace5579..27d2b0f7d 100644
--- a/mfront/tests/behaviours/LogarithmicStrainStandardElasticity3.mfront
+++ b/mfront/tests/behaviours/LogarithmicStrainStandardElasticity3.mfront
@@ -41,6 +41,6 @@
 @UpdateAuxiliaryStateVariables {
   const auto ds = eval(theta * sig + (1 - theta) * sig0 - sigel);
   if (sqrt(ds | ds) > 10 * epsilon * D(0,0)) {
-    tfel::raise("invalid stress prediction");
+    return tfel::fail("invalid stress prediction");
   }
 }
\ No newline at end of file
diff --git a/mfront/tests/behaviours/LogarithmicStrainStandardElasticity4.mfront b/mfront/tests/behaviours/LogarithmicStrainStandardElasticity4.mfront
index 097aaacb0..8624a4165 100644
--- a/mfront/tests/behaviours/LogarithmicStrainStandardElasticity4.mfront
+++ b/mfront/tests/behaviours/LogarithmicStrainStandardElasticity4.mfront
@@ -47,6 +47,6 @@
 @UpdateAuxiliaryStateVariables {
   const auto ds = eval(theta * sig + (1 - theta) * sig0 - sigel);
   if (sqrt(ds | ds) > 10 * epsilon * D(0,0)) {
-    tfel::raise("invalid stress prediction");
+    return tfel::fail("invalid stress prediction");
   }
 }
\ No newline at end of file
diff --git a/mfront/tests/behaviours/MonoCrystal_DD_CC.mfront b/mfront/tests/behaviours/MonoCrystal_DD_CC.mfront
index 444c29b3c..e2c935b16 100644
--- a/mfront/tests/behaviours/MonoCrystal_DD_CC.mfront
+++ b/mfront/tests/behaviours/MonoCrystal_DD_CC.mfront
@@ -98,7 +98,8 @@
   for(unsigned short k=0;k!=12;++k){
     omegap[k]=omega[k]+theta*domega[k];
     if(omegap[k] < 0.) {
-      throw(runtime_error("omega négatif"));
+      this->abortResolution();
+      return tfel::fail<runtime_error>("omega négatif");
     } 
   }
 
@@ -134,7 +135,8 @@
     }
       
     if(alphat[i] < 0.) {
-      throw(runtime_error("alphat[i] négatif"));
+      this->abortResolution();
+      return tfel::fail<runtime_error>("alphat[i] négatif");
     } 
     alphat[i] = sqrt(alphat[i]/omega_tot[i]) ;
     ls[i] = max ( (lambda[i] - 2.0 * alphat[i] * Rs[i]) , Lc );
@@ -152,7 +154,8 @@
     }
       
     if(tau_eff[i] > tau_0) {
-      throw(runtime_error("tau_eff[i] > tau_0"));
+      this->abortResolution();
+      return tfel::fail<runtime_error>("tau_eff[i] > tau_0");
     }
     real DG_eff = DeltaG_0 ;
     if (tau_eff[i] > 0.0) {
@@ -161,9 +164,14 @@
     gamman[i] = dt * omega_mob / b* H * ls[i] * exp ( - DG_eff / k_b / T ) * sgn_tau ;
 
     if(abs(tau_c[i])<1.e-50) {
+#ifndef TFEL_DEVICE_CODE
       ostringstream msg;
       msg << "invalid value for tau_c (" << tau_c[i] << ").";
       throw(runtime_error(msg.str()));
+#else  /* TFEL_DEVICE_CODE */
+      this->abortResolution();
+      return false;
+#endif /* TFEL_DEVICE_CODE */
     }      
     gammap[i] = dt * gamma0 * pow( ( abs(tau[i]) / tau_c[i] ), pn) * sgn_tau ;
       
diff --git a/mfront/tests/behaviours/MonoCrystal_DD_CC_Irradiation.mfront b/mfront/tests/behaviours/MonoCrystal_DD_CC_Irradiation.mfront
index 3dd022216..8e3ca1532 100644
--- a/mfront/tests/behaviours/MonoCrystal_DD_CC_Irradiation.mfront
+++ b/mfront/tests/behaviours/MonoCrystal_DD_CC_Irradiation.mfront
@@ -101,7 +101,8 @@
   for(unsigned short k=0;k!=12;++k){
     omegap[k]=omega[k]+theta*domega[k];
     if(omegap[k] < 0.) {
-      throw(runtime_error("omega négatif"));
+      this->abortResolution();
+      return tfel::fail<runtime_error>("omega négatif");
     } 
   }
 
@@ -137,7 +138,8 @@
     }
     alphat[i] += a_irr * omega_irr[i]  ;      
     if(alphat[i] < 0.) {
-      throw(runtime_error("alphat[i] négatif"));
+      this->abortResolution();
+      return tfel::fail<runtime_error>("alphat[i] négatif");
     } 
     alphat[i] = sqrt(alphat[i]/omega_tot[i]) ;
     ls[i] = max ( (lambda[i] - 2.0 * alphat[i] * Rs[i]) , Lc );
@@ -155,7 +157,8 @@
     }
       
     if(tau_eff[i] > tau_0) {
-      throw(runtime_error("tau_eff[i] > tau_0"));
+      this->abortResolution();
+      return tfel::fail<runtime_error>("tau_eff[i] > tau_0");
     }
     real DG_eff = DeltaG_0 ;
     if (tau_eff[i] > 0.0) {
@@ -164,9 +167,14 @@
     gamman[i] = dt * omega_mob / b* H * ls[i] * exp ( - DG_eff / k_b / T ) * sgn_tau ;
 
     if(abs(tau_c[i])<1.e-50) {
+#ifndef TFEL_DEVICE_CODE
       ostringstream msg;
       msg << "invalid value for tau_c (" << tau_c[i] << ")";
       throw(runtime_error(msg.str()));
+#else  /* TFEL_DEVICE_CODE */
+      this->abortResolution();
+      return false;
+#endif /* TFEL_DEVICE_CODE */
     }      
     gammap[i] = dt * gamma0 * pow( ( abs(tau[i]) / tau_c[i] ), pn) * sgn_tau ;
       
diff --git a/mfront/tests/behaviours/MonoCrystal_DD_CC_Irradiation_NumericalJacobian.mfront b/mfront/tests/behaviours/MonoCrystal_DD_CC_Irradiation_NumericalJacobian.mfront
index ebaed920e..c05f6f068 100644
--- a/mfront/tests/behaviours/MonoCrystal_DD_CC_Irradiation_NumericalJacobian.mfront
+++ b/mfront/tests/behaviours/MonoCrystal_DD_CC_Irradiation_NumericalJacobian.mfront
@@ -81,8 +81,11 @@
   for (unsigned short k = 0; k != 12; ++k) {
     omegap[k] = omega[k] + theta * domega[k];
     if (omegap[k] < 0.) {
+#ifndef TFEL_DEVICE_CODE
       cout << " k" << k << " omega " << omegap[k] << endl;
-      throw(runtime_error("omega négatif"));
+#endif /* TFEL_DEVICE_CODE */
+      this->abortResolution();
+      return tfel::fail<runtime_error>("omega négatif");
     }
   }
 
@@ -114,7 +117,8 @@
     }
     alpha_AT += a_irr * omega_irr[i];
     if (alpha_AT < 0.) {
-      throw(runtime_error("alpha_AT négatif"));
+      this->abortResolution();
+      return tfel::fail<runtime_error>("alpha_AT négatif");
     }
     alpha_AT = sqrt(alpha_AT / omega_tot);
     real ls = max((lambda - 2.0 * alpha_AT * Rs), Lc);
@@ -134,8 +138,11 @@
     }
 
     if (tau_eff > tau_0) {
+#ifndef TFEL_DEVICE_CODE
       cout << "tau_eff " << tau_eff << " tau_0 " << tau_0 << endl;
-      throw(runtime_error("tau_eff > tau_0"));
+#endif /* TFEL_DEVICE_CODE */
+      this->abortResolution();
+      return tfel::fail<runtime_error>("tau_eff > tau_0");
     }
     real DG_eff = DeltaG_0;
     if (tau_eff > 0.0) {
@@ -144,9 +151,14 @@
     real gamma_nuc = omega_mob / b * H * ls * exp(-DG_eff / k_b / T) * sgn_tau;
 
     if (abs(tau_c) < 1.e-50) {
+#ifndef TFEL_DEVICE_CODE
       ostringstream msg;
       msg << "invalid value for tau_c (" << tau_c << ")";
       throw(runtime_error(msg.str()));
+#else  /* TFEL_DEVICE_CODE */
+      this->abortResolution();
+      return false;
+#endif /* TFEL_DEVICE_CODE */
     }
     real gamma_prob = gamma0 * pow((abs(tau[i]) / tau_c), pn) * sgn_tau;
 
diff --git a/mfront/tests/behaviours/MonoCrystal_DD_CC_LevenbergMarquardt.mfront b/mfront/tests/behaviours/MonoCrystal_DD_CC_LevenbergMarquardt.mfront
index 95b5686e2..0efca4521 100644
--- a/mfront/tests/behaviours/MonoCrystal_DD_CC_LevenbergMarquardt.mfront
+++ b/mfront/tests/behaviours/MonoCrystal_DD_CC_LevenbergMarquardt.mfront
@@ -96,7 +96,8 @@
   for(unsigned short k=0;k!=12;++k){
     omegap[k]=omega[k]+theta*domega[k];
     if(omegap[k] < 0.) {
-      throw(runtime_error("omega négatif"));
+      this->abortResolution();
+      return tfel::fail<runtime_error>("omega négatif");
     } 
   }
 
@@ -132,7 +133,8 @@
     }
       
     if(alphat[i] < 0.) {
-      throw(runtime_error("alphat[i] négatif"));
+      this->abortResolution();
+      return tfel::fail<runtime_error>("alphat[i] négatif");
     } 
     alphat[i] = sqrt(alphat[i]/omega_tot[i]) ;
     ls[i] = max ( (lambda[i] - 2.0 * alphat[i] * Rs[i]) , Lc );
@@ -150,7 +152,8 @@
     }
       
     if(tau_eff[i] > tau_0) {
-      throw(runtime_error("tau_eff[i] > tau_0"));
+      this->abortResolution();
+      return tfel::fail<runtime_error>("tau_eff[i] > tau_0");
     }
     real DG_eff = DeltaG_0 ;
     if (tau_eff[i] > 0.0) {
@@ -159,9 +162,14 @@
     gamman[i] = dt * omega_mob / b* H * ls[i] * exp ( - DG_eff / k_b / T ) * sgn_tau ;
 
     if(abs(tau_c[i])<1.e-50) {
+#ifndef TFEL_DEVICE_CODE
       ostringstream msg;
       msg << "invalid value for tau_c (" << tau_c[i] << ").";
       throw(runtime_error(msg.str()));
+#else  /* TFEL_DEVICE_CODE */
+      this->abortResolution();
+      return false;
+#endif /* TFEL_DEVICE_CODE */
     }      
     gammap[i] = dt * gamma0 * pow( ( abs(tau[i]) / tau_c[i] ), pn) * sgn_tau ;
       
diff --git a/mfront/tests/behaviours/MonoCrystal_DD_CC_NumericalJacobian.mfront b/mfront/tests/behaviours/MonoCrystal_DD_CC_NumericalJacobian.mfront
index 0daec7884..ff304b7da 100644
--- a/mfront/tests/behaviours/MonoCrystal_DD_CC_NumericalJacobian.mfront
+++ b/mfront/tests/behaviours/MonoCrystal_DD_CC_NumericalJacobian.mfront
@@ -82,8 +82,11 @@ omega.setNormalisationFactor(1.e-2);
   for(unsigned short k=0;k!=12;++k){
     omegap[k]=omega[k]+theta*domega[k];
     if(omegap[k] < 0.) {
+#ifndef TFEL_DEVICE_CODE
       cout << " k" << k << " omega " << omegap[k] << endl ;
-      throw(runtime_error("omega négatif"));
+#endif /* TFEL_DEVICE_CODE */
+      this->abortResolution();
+      return tfel::fail<runtime_error>("omega négatif");
     } 
   }
 
@@ -111,7 +114,8 @@ omega.setNormalisationFactor(1.e-2);
     }
       
     if(alpha_AT < 0.) {
-      throw(runtime_error("alpha_AT négatif"));
+      this->abortResolution();
+      return tfel::fail<runtime_error>("alpha_AT négatif");
     } 
     alpha_AT = sqrt(alpha_AT/omega_tot) ;
     real ls = max ( (lambda - 2.0 * alpha_AT * Rs) , Lc );
@@ -129,8 +133,11 @@ omega.setNormalisationFactor(1.e-2);
     }
       
     if(tau_eff > tau_0) {
+#ifndef TFEL_DEVICE_CODE
       cout << "tau_eff " << tau_eff << " tau_0 " << tau_0 << endl ;
-      throw(runtime_error("tau_eff > tau_0"));
+#endif /* TFEL_DEVICE_CODE */
+      this->abortResolution();
+      return tfel::fail<runtime_error>("tau_eff > tau_0");
     }
     real DG_eff = DeltaG_0 ;
     if (tau_eff > 0.0) {
@@ -139,9 +146,14 @@ omega.setNormalisationFactor(1.e-2);
     real gamma_nuc = omega_mob / b* H * ls * exp ( - DG_eff / k_b / T ) * sgn_tau ;
 
     if(abs(tau_c)<1.e-50) {
+#ifndef TFEL_DEVICE_CODE
       ostringstream msg;
       msg << "invalid value for tau_c (" << tau_c << ").";
       throw(runtime_error(msg.str()));
+#else  /* TFEL_DEVICE_CODE */
+      this->abortResolution();
+      return false;
+#endif /* TFEL_DEVICE_CODE */
     }      
     real gamma_prob = gamma0 * pow( ( abs(tau[i]) / tau_c ), pn) * sgn_tau ;
       
diff --git a/mfront/tests/behaviours/OrthotropicElastic6.mfront b/mfront/tests/behaviours/OrthotropicElastic6.mfront
index 6ab95d784..3a13ee74b 100644
--- a/mfront/tests/behaviours/OrthotropicElastic6.mfront
+++ b/mfront/tests/behaviours/OrthotropicElastic6.mfront
@@ -69,8 +69,10 @@ AxisymmetricalGeneralisedPlaneStress::sigzz.setGlossaryName("AxialStress");
     Dt(0,2)=D_tdt(0,2)-D_tdt(0,1)*D_tdt(2,1)/D_tdt(1,1);
     Dt(2,2)=D_tdt(2,2)-D_tdt(1,2)*D_tdt(2,1)/D_tdt(1,1);
     Dt(0,1)=Dt(1,0)=Dt(1,2)=Dt(2,1)=Dt(1,1)=stress(0);
+#ifndef TFEL_DEVICE_CODE
     std::cout << "Dt: " << Dt << '\n';
     std::cout << "Dt: " << Dt(1,2) << '\n';
+#endif /* TFEL_DEVICE_CODE */
 } else {
     Dt = D_tdt;
   }
diff --git a/mfront/tests/behaviours/ParameterTest.mfront b/mfront/tests/behaviours/ParameterTest.mfront
index 702e02aa5..7b86f01c3 100644
--- a/mfront/tests/behaviours/ParameterTest.mfront
+++ b/mfront/tests/behaviours/ParameterTest.mfront
@@ -21,12 +21,14 @@ nu.setGlossaryName("PoissonRatio");
   constexpr const auto eps = real(1.e-12);
   const auto dy = young-stress(100e9);
   const auto dn = nu-real(0.2);
-  raise_if(abs(dy)>eps*young,
-	   "invalid value for the "
-	   "Young modulus");
-  raise_if(abs(dn)>eps*nu,
-	   "invalid value for the "
-	   "Poisson ratio");
+  if (abs(dy) > eps * young) {
+    return tfel::fail("invalid value for the "
+                      "Young modulus");
+  }
+  if (abs(dn) > eps * nu) {
+    return tfel::fail("invalid value for the "
+                      "Poisson ratio");
+  }
   lambda = computeLambda(young,nu);
   mu = computeMu(young,nu);
 }
diff --git a/mfront/tests/behaviours/ParameterTest2.mfront b/mfront/tests/behaviours/ParameterTest2.mfront
index 2235642fd..fc83e1f45 100644
--- a/mfront/tests/behaviours/ParameterTest2.mfront
+++ b/mfront/tests/behaviours/ParameterTest2.mfront
@@ -21,12 +21,14 @@ a.setEntryName("ArrayOfParameters");
 @LocalVariable stress lambda,mu;
 
 @InitLocalVariables{
-  raise_if(std::abs(a[0]-15)>1.e12,
-	   "invalid value for the "
-	   "Young modulus");
-  raise_if(std::abs(a[1]+3)>1.e12,
-	   "invalid value for the "
-	   "Young modulus");
+  if (std::abs(a[0] - 15) > 1.e12) {
+    return tfel::fail("invalid value for the "
+                      "Young modulus");
+  }
+  if (std::abs(a[1] + 3) > 1.e12) {
+    return tfel::fail("invalid value for the "
+                      "Young modulus");
+  }
   lambda = computeLambda(young,nu);
   mu = computeMu(young,nu);
 }
diff --git a/mfront/tests/behaviours/SaintVenantKirchhoffElasticity2.mfront b/mfront/tests/behaviours/SaintVenantKirchhoffElasticity2.mfront
index 4a4970677..2286ae3d2 100644
--- a/mfront/tests/behaviours/SaintVenantKirchhoffElasticity2.mfront
+++ b/mfront/tests/behaviours/SaintVenantKirchhoffElasticity2.mfront
@@ -72,8 +72,8 @@ nu.setGlossaryName("PoissonRatio");
       Dt = convert<FiniteStrainBehaviourTangentOperatorBase::ABAQUS,
 		   FiniteStrainBehaviourTangentOperatorBase::SPATIAL_MODULI,N,stress>(Cs,F0,F1,sig);
     } else {
-      throw(runtime_error("SaintVenantKirchhoffElasticity : "
-			  "unsupported tangent operator"));
+      return tfel::fail<runtime_error>("SaintVenantKirchhoffElasticity : "
+                                       "unsupported tangent operator");
     }
   }
 }
diff --git a/mfront/tests/behaviours/SaintVenantKirchhoffElasticityWithSubStepping.mfront b/mfront/tests/behaviours/SaintVenantKirchhoffElasticityWithSubStepping.mfront
index 9e8076435..39eb28b9a 100644
--- a/mfront/tests/behaviours/SaintVenantKirchhoffElasticityWithSubStepping.mfront
+++ b/mfront/tests/behaviours/SaintVenantKirchhoffElasticityWithSubStepping.mfront
@@ -22,7 +22,7 @@ nu.setGlossaryName("PoissonRatio");
 
 @InitLocalVariables{
   if(dt>0.025){
-    throw(DivergenceException("invalid time step"));
+    return tfel::fail<DivergenceException>("invalid time step");
   }
   lambda = computeLambda(young,nu);
   mu     = computeMu(young,nu);
diff --git a/mfront/tests/behaviours/StandardElasticity/TestComputeElasticPrediction.mfront b/mfront/tests/behaviours/StandardElasticity/TestComputeElasticPrediction.mfront
index ae7c76190..5bff0b7c0 100644
--- a/mfront/tests/behaviours/StandardElasticity/TestComputeElasticPrediction.mfront
+++ b/mfront/tests/behaviours/StandardElasticity/TestComputeElasticPrediction.mfront
@@ -21,7 +21,7 @@
 
 @UpdateAuxiliaryStateVariables{
   if(abs(sigel-sig)>10*std::numeric_limits<real>::epsilon()*young){
-    throw(runtime_error("TestComputeElasticPrediction: "
-			"computeElasticPrediction failed"));
+    return tfel::fail<runtime_error>("TestComputeElasticPrediction: "
+                                     "computeElasticPrediction failed");
   }
 }
diff --git a/mfront/tests/behaviours/StandardElasticity/TestComputeElasticPrediction2.mfront b/mfront/tests/behaviours/StandardElasticity/TestComputeElasticPrediction2.mfront
index 2cfca5f32..76ab5da30 100644
--- a/mfront/tests/behaviours/StandardElasticity/TestComputeElasticPrediction2.mfront
+++ b/mfront/tests/behaviours/StandardElasticity/TestComputeElasticPrediction2.mfront
@@ -29,7 +29,7 @@
 
 @UpdateAuxiliaryStateVariables{
   if(abs(sigel-sig)>10*numeric_limits<real>::epsilon()*D(0,0)){
-    throw(runtime_error("TestComputeElasticPrediction: "
-			"computeElasticPrediction failed"));
+    return tfel::fail<runtime_error>("TestComputeElasticPrediction: "
+                                     "computeElasticPrediction failed");
   }
 }
diff --git a/mfront/tests/behaviours/TemperatureUpdateOnSubSteppingCheck.mfront b/mfront/tests/behaviours/TemperatureUpdateOnSubSteppingCheck.mfront
index 875e9a6a5..c1307e373 100644
--- a/mfront/tests/behaviours/TemperatureUpdateOnSubSteppingCheck.mfront
+++ b/mfront/tests/behaviours/TemperatureUpdateOnSubSteppingCheck.mfront
@@ -18,7 +18,7 @@
 
 @InitLocalVariables{
   if(dt>1){
-    throw(DivergenceException());
+    return tfel::fail<DivergenceException>();
   }
   constexpr const auto e  = numeric_limits<real>::epsilon();
   constexpr const auto Ti = 293.15;
@@ -26,7 +26,7 @@
   constexpr const auto te = 3600;
   const auto Tt = Ti+(Te-Ti)*(ct/te);
   if(std::abs(T-Tt)>100*Te*e){
-    throw(runtime_error("invalid temperature value"));
+    return tfel::fail<runtime_error>("invalid temperature value");
   }
 }
 
diff --git a/mfront/tests/models/SiC_IrradiationSwellingModel_GoFaster.mfront b/mfront/tests/models/SiC_IrradiationSwellingModel_GoFaster.mfront
index 47c45d502..7a6548c51 100644
--- a/mfront/tests/models/SiC_IrradiationSwellingModel_GoFaster.mfront
+++ b/mfront/tests/models/SiC_IrradiationSwellingModel_GoFaster.mfront
@@ -43,10 +43,14 @@ n.setDepth(1);
   if((0<=theta)&&(theta<=1500)){
     swellingSaturation = 0.04*exp(-theta/370);
   } else {
+#ifndef TFEL_DEVICE_CODE
     ostringstream msg;
     msg << "SIC_SwellingModel::computeSwellingIncrement : ";
     msg << "temperature is out of bounds (" << theta << ")";
     throw(runtime_error(msg.str()));
+#else  /* TFEL_DEVICE_CODE */
+    return false;
+#endif /* TFEL_DEVICE_CODE */
   }
   incr  = swellingSaturation/Phi0;
   incr *= exp(-Phi/Phi0);
-- 
2.43.0

